README.md
Overview
========

[![Build Status](https://travis-ci.org/bruth/django-tracking2.svg?branch=master)](https://travis-ci.org/bruth/django-tracking2)
[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/bruth/django-tracking2/trend.png)](https://bitdeli.com/free "Bitdeli Badge")

django-tracking2 tracks the length of time visitors and registered users
spend on your site. Although this will work for websites, this is more
applicable to web _applications_ with registered users. This does
not replace (nor intend) to replace client-side analytics which is
great for understanding aggregate flow of page views.

**Note: This is not a new version of [django-tracking]. These apps
have very different approaches and, ultimately, goals of tracking users.
This app is about keeping a history of visitor sessions, rather than the
current state of the visitor.**

[django-tracking]: https://github.com/codekoala/django-tracking

Requirements
============
* Django's [session framework][1] installed
* [South][2] (if you want to use the packaged migrations)

[1]: https://docs.djangoproject.com/en/1.3/topics/http/sessions/
[2]: http://pypi.python.org/pypi/South

Download
========
```bash
pip install django-tracking2
```

Setup
=====
Add `tracking` to your project's `INSTALLED_APPS` setting:

```python
INSTALLED_APPS = (
    ...
    'tracking',
    ...
)
```

If you use Django 1.7 `tracking` app should follow the app with your user model

Add `tracking.middleware.VisitorTrackingMiddleware` to your project's
`MIDDLEWARE_CLASSES` before the `SessionMiddleware`:

```python
MIDDLEWARE_CLASSES = (
    ...
    'tracking.middleware.VisitorTrackingMiddleware',
    'django.contrib.sessions.middleware.SessionMiddleware',
    ...
)
```

Django 1.7
----------
Django 1.7 brings changes to the way database migrations are handled. If
you do not use database migrations you should not be worried. If you use
south and have not upgraded to django 1.7, you'll have to upgrade to
south==1.0.

With south 1.0, south-based migrations are found in the `south_migrations`
directory while django 1.7+ migrations are in the `migrations` directory.

Settings
--------
`TRACK_AJAX_REQUESTS` - If True, AJAX requests will be tracked. Default
is False

`TRACK_ANONYMOUS_USERS` - If False, anonymous users will not be tracked.
Default is True

`TRACK_PAGEVIEWS` - If True, individual pageviews will be tracked.

`TRACK_IGNORE_URLS` - A list of regular expressions that will be matched
against the `request.path_info` (`request.path` is stored, but not matched
against). If they are matched, the visitor (and pageview) record will not
be saved. Default includes 'favicon.ico' and 'robots.txt'. Note, static and
media are not included since they should be served up statically Django's
static serve view or via a lightweight server in production. Read more
[here](https://docs.djangoproject.com/en/dev/howto/static-files/#serving-other-directories)

`TRACK_IGNORE_STATUS_CODES` - A list of HttpResponse status codes that will be ignored.
If the HttpResponse object has a `status_code` in this blacklist, the pageview record 
will not be saved. For example,

```python
TRACK_IGNORE_STATUS_CODES = [400, 404, 403, 405, 410, 500]
```

`TRACK_REFERER` - If True, referring site for all pageviews will be tracked.  Default is False

`TRACK_QUERY_STRING` - If True, query string for all pageviews will be tracked.  Default is False

Views
-----
To view aggregate data about all visitors and per-registered user stats,
do the following:

Include `tracking.urls` in your `urls.py`:

```python
urlpatterns = patterns('',
    ...
    url(r'^tracking/', include('tracking.urls')),
    ...
)
```

These urls are protected by a custom Django permission `tracking.view_visitor`.
Thus only superusers and users granted this permission can view these pages.

Available URLs
--------------
* `/` - overview of all visitor activity, includes a time picker for
        filtering.

Templates
---------
* `tracking/dashboard.html` - for the dashboard page
* `tracking/snippets/stats.html` - standalone content for the dashboard page
  (simplifies overriding templates)


CHANGELOG
2015-07-22	Michael J. Schultz	<mjschultz@gmail.com>

	* Add continuous integration (CI) tests for Django 1.8
	* Drop CI tests for Django 1.5 and 1.6

2015-07-10  Michael J. Schultz  <mjschultz@gmail.com>

	* Handle race condition when saving visitor in middleware.

2014-12-27	Byron Ruth	<b@devel.io>

	* Replace IP validation with Django's built-in validators. This also
	removes a Windows-specific issue of not having the `socket.inet_aton`
	function. (#48, #47)

2014-12-08  Michael J. Schultz  <mjschultz@gmail.com>

    * Add tests to existing codebase with minor bugfixes
    * Change /dashboard/ view to include past 7 days by default (#20, #28)
    * Improve timezone support by removing `.date()` calls (#39)
    * Increase scope of `TRACK_IGNORE_URLS` beyond Pageview tracking (#45)

2013-11-07  Byron Ruth  <b@devel.io>

    * Add PageView.method field for tracking the request method used

2013-11-06  Byron Ruth  <b@devel.io>

    * Update deprecated import path for Django 1.6+

2013-11-01  Byron Ruth  <b@devel.io>

    * Move stats dashboard endpoint / rather than /dashboard/
    * Update initial migration to support custom user models for Django 1.5+

2013-09-30  Byron Ruth  <b@devel.io>

    * Fix MANIFEST.in file to graft templates directory

2013-09-08	Byron Ruth	<b@devel.io>

    * Add support custom user model in Django 1.5+

2013-07-14	Byron Ruth	<b@devel.io>

    * Remove deprecated module import for GIS by Tarak Blah
    * Added start and end date to stats view context by Tarak Blah
    * Remove registered users as requirement in manager by Tarak Blah

2013-03-16  Byron Ruth  <b@devel.io>

    * Fix offset-aware and offset-naive datetime errors by Chris Franklin

2012-03-14  Byron Ruth  <b@devel.io>

    * Add internal cache mechanism to reduce database hits in middleware
    by one. Cache invalidates after `SESSION_COOKIE_AGE` and when the user
    explicitly logs out.

2012-03-08  Byron Ruth  <b@devel.io>

    * Add `TRACK_PAGEVIEWS` and `TRACK_IGNORE_URLS` settings.
    * Deprecate `TRACKING_USE_GEOIP` in favor of `TRACK_USING_GEOIP`
    (even though this is not an officially supported feature)

2012-03-03  Byron Ruth  <b@devel.io>

    * Add `TRACK_ANONYMOUS_USERS` setting

2012-01-30  Byron Ruth  <b@devel.io>

    * Fix bug where the 'User-Agent' header was assumed to
    exist which caused an exception.

2012-01-25  Byron Ruth  <b@devel.io>

    * Add `TRACK_AJAX_REQUESTS` setting

2012-01-18  Byron Ruth  <b@devel.io>

    * Initial release


tracking/views.py
import logging

from datetime import timedelta

from django import forms
from django.shortcuts import render
from django.contrib.auth.decorators import permission_required
from django.utils.timezone import now

from tracking.models import Visitor, Pageview
from tracking.settings import TRACK_PAGEVIEWS

log = logging.getLogger(__file__)

# tracking wants to accept more formats than default, here they are
input_formats = [
    '%Y-%m-%d %H:%M:%S',    # '2006-10-25 14:30:59'
    '%Y-%m-%d %H:%M',       # '2006-10-25 14:30'
    '%Y-%m-%d',             # '2006-10-25'
    '%Y-%m',                # '2006-10'
    '%Y',                   # '2006'
]


class DashboardForm(forms.Form):
    start = forms.DateTimeField(required=False, input_formats=input_formats)
    end = forms.DateTimeField(required=False, input_formats=input_formats)


@permission_required('tracking.view_visitor')
def dashboard(request):
    "Counts, aggregations and more!"
    end_time = now()
    start_time = end_time - timedelta(days=7)
    defaults = {'start': start_time, 'end': end_time}

    form = DashboardForm(data=request.GET or defaults)
    if form.is_valid():
        start_time = form.cleaned_data['start']
        end_time = form.cleaned_data['end']

    # determine when tracking began
    try:
        obj = Visitor.objects.order_by('start_time')[0]
        track_start_time = obj.start_time
    except (IndexError, Visitor.DoesNotExist):
        track_start_time = now()

    # If the start_date is before tracking began, warn about incomplete data
    warn_incomplete = (start_time < track_start_time)

    # queries take `date` objects (for now)
    user_stats = Visitor.objects.user_stats(start_time, end_time)
    visitor_stats = Visitor.objects.stats(start_time, end_time)
    if TRACK_PAGEVIEWS:
        pageview_stats = Pageview.objects.stats(start_time, end_time)
    else:
        pageview_stats = None

    context = {
        'form': form,
        'track_start_time': track_start_time,
        'warn_incomplete': warn_incomplete,
        'user_stats': user_stats,
        'visitor_stats': visitor_stats,
        'pageview_stats': pageview_stats,
    }
    return render(request, 'tracking/dashboard.html', context)
